### DISCUSSION TITLE
Wemos/Lolin S2 Mini w/ESP32-S2, D1 Mini pin compatible

### DISCUSSION THREAD

**Discussion by durapensa** [arendst/Tasmota#12926] [Show and tell]
I ordered the $4 Wemos/Lolin S2 Mini to give it a try with [pre-loaded] MicroPython. I'll build & test Tasmota using the [ESP32-S2 board definition](https://github.com/arendst/Tasmota/blob/development/boards/esp32s2.json).

More on this board at the [wemos.cc wiki](https://www.wemos.cc/en/latest/s2/s2_mini.html) & the [CNX Software writeup](https://www.cnx-software.com/2021/08/18/lolin-s2-mini-esp32-s2-wifi-iot-board-works-with-wemos-d1-mini-shields/).

**Comment by barbudor** [comment]
I think ESP32-S2 are a great stuff for CircuitPython because CircuitPython mount the internal flash as a mass storage and you can easily edit or drop *.py file directly.
For MicroPython, this doesn't bring much value over a standard ESP32
For Tasmota, ESP32-C3 is more interresting as even cheaper and a good replacement for ESP8266

**Reply by durapensa** [reply]
I plan to try some existing ESP32-S2 CircuitPython builds with this board, and likely a new board definition for CP will be trivial. I'd also like to try to load a UF2 bootloader (something the ESP32 cannot do, to my knowledge).

Even for MicroPython, the advantage of ESP32-S2 over ESP32 is price - it's a cheaper chip to produce and seems it or the ESP32-C3 may be the high-volume successor to ESP8266.  Without knowing Espressif's constraints for producing these two chips (if those constraints differ at all), I guess only time & market will decide this one.

Perhaps Wemos/Lolin will make a C3 Mini too - I don't know why they chose ESP32-S2 over ESP32-C3.

For this Wemos S2 Mini, getting Tasmota's Berry scripting on a $4 D1 Mini form factor is a big boon.

**Reply by s-hadinger** [reply]
Take a look at Berry scripting. It's similar to Python but much lighter in resources needs

**Reply by durapensa** [reply]
> Take a look at Berry scripting. It's similar to Python but much lighter in resources needs

See the last sentence in the block you replied to. 🙂

**Reply by sfromis** [reply]
One reason for not choosing the C3 is that it is very new in the market, only recently out of prototype stage, and some people may not want to do the bleeding on the leading edge. Outside of potential philosophical preferences for Risc-V, the advantages of C3 are small, and production delta cost may not be much of a factor in how Espressif choose to price chips.

**Comment by Jason2866** [comment]
S2 Arduino has a bug with flash read/write which makes LittleFS to fail. See https://github.com/espressif/arduino-esp32/issues/5550
There is no need to wait for Lolin to do C3 boards. There are already nice ones from others

**Reply by durapensa** [reply]
> There is no need to wait for Lolin to do C3 boards. There are already nice ones from others

Of course, but I want something that fits the small project boxes & shields I'm already using. A drop-in replacement for D1 Mini would be perfect and I'd snap up a C3 Mini in a heartbeat.

**Reply by sfromis** [reply]
Those wider (about 4 mm extra) boards similar to D1 Mini may not fit in project boxes a good fit for the ESP-12 based variant.

**Reply by durapensa** [reply]
I scoped out the LILYGO T-OI PLUS ESP32-C3 a few weeks back and really liked what I saw. Perfect for battery operation, but fewer pins broken out than the S2 Mini (assuming Lolin would follow S2 Mini design for a C3 Mini).

**Reply by sfromis** [reply]
As the ESP32-C3 chip has a lot less pins, and an ESP-12 like module is available, it would be pretty simple to make a board like the "narrow" breadboard-compatible D1 Mini. ESP32-C3 seems to be scoped as a replacement for ESP8266, not for "full" ESP32.

**Reply by barbudor** [reply]
>... it would be pretty simple to make a board like the "narrow" breadboard-compatible D1 Mini

Don't ask, they already made it
With Ch340 : 
€ 2,09 | CH340G USB-TTL converter for ESP8266 -ESP07 / ESP12 to WIFI D1 mini, with pin-headers set.
https://a.aliexpress.com/_ujCon0

Without Ch340, better for battery powered 
€ 1,36 | RobotDyn WIFI D1 mini - ESP-07 Shield, with pin-headers set compatible with Arduino boards WI-FI ESP8266 for DIY Electronic
https://a.aliexpress.com/_uIdnMA

**Reply by durapensa** [reply]
I've seen people do these module replacements. A nice thing about the S2 Mini is that it uses a bare ESP32-S2FN4R2 SoC (all flash & PSRAM & USB controller in-package) rather than a module, so smaller with fewer external components & more pins.

**Reply by blakadder** [reply]
There already is a small (but taller) footprint C3 dev board - https://t.ly/VPyw

**Reply by barbudor** [reply]
A pity they have soldered the M module on it instead of putting an edge-connector

**Reply by durapensa** [reply]
I had my eye on that C3 engineering sample dev board & its [specs](https://docs.ai-thinker.com/_media/esp32/docs/esp-c3-01m-kit-v1.0_specification.pdf). 

They also make a board with ESP-C3-12F module (soldered, no edge connector), [ESP-C3-12F-KIT](https://www.aliexpress.com/item/1005002982801980.html?spm=a2g0o.productlist.0.0.1dc36f1ezKFZh5) ([specs](https://docs.ai-thinker.com/_media/esp32/docs/esp-c3-12f-kit-v1.0_specification.pdf)).

**Comment by Jason2866** [comment]
The ESP32-S2 wemos board do not have a usb-serial converter on board! For flashing the chip USB port is used. Flashing is not possible without some (inter)action. I do not like this with a dev board. Imho a cheap design.

**Reply by durapensa** [reply]
From [Espressif docs](https://docs.espressif.com/projects/esp-idf/en/release-v4.3/esp32s2/api-guides/usb-console.html):

> On chips with an integrated USB peripheral, it is possible to use USB Communication Device Class (CDC) to implement the serial console, instead of using UART with an external USB-UART bridge chip. ESP32-S2 ROM code contains a USB CDC implementation, which supports for some basic functionality without requiring the application to include the USB stack:
> * Bidirectional serial console, which can be used with IDF Monitor or another serial monitor
> * Flashing using esptool.py and idf.py flash.
> * Device Firmware Update (DFU) interface for flashing the device using dfu-util and idf.py dfu.

RP2040 dev boards work in a similar way, plus they have a UF2 bootloader in ROM. I think this is the direction that new MCUs are headed as it becomes possible to integrate more peripherals into them. I like this approach as it increases flexibility & improves things like power draw for boards that blur the lines between dev & quasi-production.

**Reply by barbudor** [reply]
I also do prefer this approach.
Separated USB serial adapter hooked on MCU's UART with request to drive some pins at start to enable the bootloader is like 20's century technology.
This is much more flexible 
- do not block UART is you need it for something else
- can have multiple interface over USB deeply integrated with the MCU like serial + (real) debug + mass storage to ease access for file + ...

UF2 is great

**Reply by durapensa** [reply]
I'll experiment with flashing Adafruit's [TinyUF2](https://github.com/adafruit/tinyuf2/tree/master/ports/esp32s2) to the S2 Mini.

CircuitPython and its libs will be easier to load, but I suspect that firmware images that are unaware of TinyUF2 (probably including Tasmota) will clobber TinyUF2 when flashed. I'll test whether [uf2conv](https://github.com/microsoft/uf2/blob/master/utils/uf2conv.md) is usable in this regard to convert a built tasmota32s2.bin to UF2 (I suspect that partitions will collide, but can only try).

**Reply by s-hadinger** [reply]
Firmware image is loaded in `ota_x` partitions and don't change bootloader nor uf2 partition. So you should be safe from further updates. 

Of course if you flash via Vscode it will overwrite the partition definition

**Reply by Jason2866** [reply]
with UF2, you loose 256k flash space. You may have problems with Tasmota...

**Reply by durapensa** [reply]
Okay, nice (I don't use Vscode). I see from [tinyuf2/.../esp32s2/README.md](https://github.com/adafruit/tinyuf2/blob/master/ports/esp32s2/README.md):

> Following is typical partition for 4MB flash, check out the `partition-xMB.csv` for details.

```
# Name,   Type, SubType, Offset,  Size, Flags
# bootloader.bin,,          0x1000, 32K
# partition table,          0x8000, 4K

nvs,      data, nvs,      0x9000,  20K,
otadata,  data, ota,      0xe000,  8K,
ota_0,    0,    ota_0,   0x10000,  1408K,
ota_1,    0,    ota_1,  0x170000,  1408K,
uf2,      app,  factory,0x2d0000,  256K,
ffat,     data, fat,    0x310000,  960K,
```

And

> TinyUF2 bootloader will always write/update firmware to ota_0 partition...After complete writing, uf2 will set the ota0 as bootable and reset, and the application should be running in the next boot. 

The S2 Mini is most similar to [unexpectedmaker_tinys2](https://github.com/adafruit/tinyuf2/blob/master/ports/esp32s2/boards/unexpectedmaker_tinys2/board.h) (same ESP32-S2FN4R2 SoC, and I'll wire up a WS2812B NeoPixel), so this all seems to indicate that after running `uf2conv.py` on `tasmota32s2.bin` and writing the resultant `tasmota32s2.UF2` to UF2_VOLUME  `TINYS2BOOT` (from unexpectedmaker_tinys2/board.h linked above), Tasmota OTA should continue to work fine in conjunction with TinyUF2?

**Reply by Jason2866** [reply]
You have to change the `partition-xMB.csv` table. Tasmota needs a SPIFFS partition, since LittleFS is enabled in all builds >1MB flash.

**Reply by Jason2866** [reply]
When the Arduino flash bug for the S2 is solved, i will provide a Tasmota Platformio setup for S2 with TinyUF2 too

**Reply by Jason2866** [reply]
Had a look into tinyuf2. It makes no sense in combination with Tasmota. To get Tasmota workin a different partition setup is needed. For this the complete! flash needs to be reflashed. If Tasmota is running updates are done via OTA

**Reply by durapensa** [reply]
For a normal Tasmota install, I agree that TinyUF2 is unnecessary. Getting Tasmota installation to play nice with boards that have TinyUF2 factory-installed (in ROM or in flash) would be nice. In the case of tinkering with this new S2 Mini, I'd like the option of using TinyUF2 & switching firmware without erase or complete flash, if all partitions can fit & co-exist in 4MB flash.

Looking at `platformio-core/Tasmota/esp32_partition_app1572k_spiffs983k.csv` from a recent build, 1572k is the smallest ota_x size I found, and `platformio-core/Tasmota/build_output/firmware/tasmota32.bin` is 1244k. I'm wondering if the 1408k size example from the TinyUF2 docs can be reduced, or if I should shave off from the 960k FAT partition to make room for a SPIFFS partition and whether this has a minimum size requirement too.

**Reply by Jason2866** [reply]
Partitions can be defined, easy. So we did for Tasmota. The problem with UF2 is you have to use the predefined ones (factory flashed). If you need a different scheme you have to reflash (the usual old way). I stopped at this point for Tasmota...
Is it possible to reduce the FAT partition? If not this would be the next killer point for bigger projects. Loosing 256k (uf2 loader) and 960k (FAT)! For a program that does more than "Hello World" a no go with a device with 4MB flash. 
The 1244k size of Tasmota is only reached by using our own Tasmota Arduino core. Using the standard core will need 245k more of flash. Imho anything lower 1500k free for program is a design fail 

Long story short: Tasmota will not run with uf2 on a 4MB device

**Reply by Jason2866** [reply]
For the S2 it makes no sense at all to use uf2. The inbuilt DFU mode works without problems. Just the manual steps are not comfortable. Windows setup (installing driver) is the tricky part. Many people will fail here. Espressif describes this badly and the mandantory part is well hidden.

**Reply by durapensa** [reply]
I'm not suggesting that UF2 be a default installation method or built-in for Tasmota for ESP32-S2 boards, only that a method to install Tasmota on UF2-enabled boards, without wiping out UF2 or other partitions, be documented (I'll document it). Some ESP32-S2 boards come with UF2 factory-installed in flash, particularly the CircuitPython focused boards (also, it's CP that uses the FAT partition), and I'll install UF2 on an S2 Mini to tinker with it.

A workaround for the FAT/SPIFFS partitioning issue could be to distribute a tasmota32s2.uf2 without LittleFS and with a note that users wanting LittleFS must perform a complete/erase DFU flash to install a full tasmota32s2.bin. Likewise, Tasmotizer could be tweaked to provide these options too upon detecting a board with UF2 in a flash partition.

**Reply by Jason2866** [reply]
Why? Everything is easy to reflash. Doing a special (crippled) build will make more issues than it solves.
Since Tasmota is Open Source maybe someone is doing it. I think Theo will not be a fan of such a special version.

**Comment by Jason2866** [comment]
We do not disagree. But for a development board i want the easiest way to program.
